modules/core/src/main/java/org/apache/ignite/internal/processors/query/RunningQueryManager.java

```diff
@@ -102,20 +139,41 @@ public Long register(String qry, GridCacheQueryType qryType, String schemaName,
      * Unregister running query.
      *
      * @param qryId Query id.
-     * @param failed {@code true} In case query was failed.
+     * @param failReason exception that
      */
-    public void unregister(Long qryId, boolean failed) {
+    public void unregister(Long qryId, @Nullable Throwable failReason) {
         if (qryId == null)
             return;
 
+        boolean failed = failReason != null;
+
         GridRunningQueryInfo qry = runs.remove(qryId);
 
-        //We need to collect query history only for SQL queries.
-        if (qry != null && isSqlQuery(qry)) {
+        // Attempt to unregister query twice.
```

Comment:
potentially it could be not exists query

Reply:
Hm, Could you please give a flow, in which we are able to unregister unknown query?
Now I still see that every register() is paired by unregister() method call, so qry == null means we've already unregistered that query.

Reply:
you are right